acceptancetests/jujupy/stream_server.py

```diff
@@ -0,0 +1,293 @@
+import socket
+import subprocess
+import tarfile
+
+from contextlib import contextmanager
+
+
+__metaclass__ = type
+
+
+log = logging.getLogger(__name__)
+
+
+class _JujuStreamData:
+    """Models stream metadata. Best used via StreamServer."""
+
+    def __init__(self, working_dir):
+        """Models Juju product simplestream metadata.
+
+        :param working_dir: Directory in which to copy agent tarballs and
+          generate stream json.
+        """
+        self.products = []
+
+        self._agent_path = os.path.join(working_dir, 'agent')
+        self._stream_path = working_dir
+
+        os.makedirs(self._agent_path)
+
+    def add_product(self, content_id, version, arch, series, agent_tgz_path):
+        shutil.copy(agent_tgz_path, self._agent_path)
+        product_dict = _generate_product_json(
+            content_id, version, arch, series, agent_tgz_path)
+        self.products.append(product_dict)
+
+    def generate_stream_data(self):
+        """Generate metadata from added products into working dir."""
+        # Late import as simplestreams.log overwrites logging handlers.
+        from simplestreams.json2streams import (
+            dict_to_item,
+            write_juju_streams
+        )
+        from simplestreams.generate_simplestreams import items2content_trees
+        # The following has been cribbed from simlestreams.json2streams.
+        # Doing so saves the need to create json files to then shell out to
+        # read those files into memory to generate the resulting json files.
+        items = (dict_to_item(item.copy()) for item in self.products)
+        updated = datetime.utcnow().strftime(
+            '%a, %d %b %Y %H:%M:%S +0000')
+        data = {'updated': updated, 'datatype': 'content-download'}
+        trees = items2content_trees(items, data)
+        return write_juju_streams(self._stream_path, trees, updated)
+
+
+class StreamServer:
+    """Provide service to create stream metadata and to serve it."""
+
+    def __init__(self, base_dir, stream_data_type=_JujuStreamData):
+        self.base_dir = base_dir
+        self.stream_data = stream_data_type(base_dir)
+
+    def add_product(self, content_id, version, arch, series, agent_tgz_path):
+        """Add a new product to generate stream data for.
+
+        :param content_id: String ID (e.g.'proposed', 'release')
+        :param version: Juju version string for product (i.e. '2.3.3')
+        :param arch: Architecture string of this product (e.g. 's390x','amd64')
+        :param series: Series string that appears in item_name
+          (e.g. 'xenial', 'centos')
+        :param agent_tgz_path: String full path to agent tarball file to use.
+          This file is copied into the JujuStreamData working dir to be served
+          up at a later date.
+        """
+        self.stream_data.add_product(
+            content_id, version, arch, series, agent_tgz_path)
+        # Re-generate when adding a product allows updating the server while
```

Comment:
If we feel the need to use this, perhaps add an optional default param set to false in the function call?